src/cpp/core/FileLock.cpp

```diff
@@ -15,163 +15,144 @@
 }
 
-Error FileLock::acquire(const FilePath& lockFilePath)
-{
-   using namespace boost::interprocess;
-
-   // make sure the lock file exists
-   if (!lockFilePath.exists())
-   {
-      Error error = core::writeStringToFile(lockFilePath, "");
-      if (error)
-         return error;
-   }
+namespace {
 
-   // try to acquire the lock
+void schedulePeriodicExecution(
+      boost::asio::deadline_timer* pTimer,
+      boost::posix_time::seconds interval,
+      boost::function<void()> callback)
+{
    try
    {
-      file_lock lock(string_utils::utf8ToSystem(lockFilePath.absolutePath()).c_str());
+      // execute callback
+      callback();
 
-      if (lock.try_lock())
+      // reschedule
+      boost::system::error_code errc;
+      pTimer->expires_at(pTimer->expires_at() + interval, errc);
+      if (errc)
       {
-         // set members
-         pImpl_->lockFilePath = lockFilePath;
-         pImpl_->lock.swap(lock);
-
-         return Success();
-      }
-      else
-      {
-         return systemError(boost::system::errc::no_lock_available,
-                            ERROR_LOCATION);
+         LOG_ERROR(Error(errc, ERROR_LOCATION));
+         return;
       }
+      
+      pTimer->async_wait(boost::bind(
+                            schedulePeriodicExecution,
+                            pTimer,
+                            interval,
+                            callback));
    }
-   catch(interprocess_exception& e)
+   catch (...)
    {
-      Error error(ec_from_exception(e), ERROR_LOCATION);
-      error.addProperty("lock-file", lockFilePath);
-      return error;
+      // swallow errors
    }
-
-   return Success();
 }
 
-Error FileLock::release()
-{
-   using namespace boost::interprocess;
-
-   // make sure the lock file exists
-   if (!pImpl_->lockFilePath.exists())
-   {
-      return systemError(boost::system::errc::no_lock_available,
-                         ERROR_LOCATION);
-   }
-
-   // always cleanup the lock file on exit
-   FilePath lockFilePath = pImpl_->lockFilePath;
-   BOOST_SCOPE_EXIT( (&lockFilePath) )
-   {
-      Error error = lockFilePath.remove();
-      if (error)
-         LOG_ERROR(error);
-   }
-   BOOST_SCOPE_EXIT_END
-
-   // try to unlock it
-   try
-   {
-      pImpl_->lock.unlock();
-      pImpl_->lock = file_lock();
-      pImpl_->lockFilePath = FilePath();
-      return Success();
-   }
-   catch(interprocess_exception& e)
-   {
-      Error error(ec_from_exception(e), ERROR_LOCATION);
-      error.addProperty("lock-file", pImpl_->lockFilePath);
-      return error;
-   }
-
-   return Success();
-}
+} // end anonymous namespace
 
-FilePath FileLock::lockFilePath() const
+void FileLock::refreshPeriodically(boost::asio::io_service& service,
+                                   boost::posix_time::seconds interval)
 {
-   return pImpl_->lockFilePath;
+   // protect against re-entrancy
+   static bool s_isRefreshing = false;
+   if (s_isRefreshing)
+      return;
+   s_isRefreshing = true;
```

Comment:
How does this ever get set back to `false`? (doesn't the static initialization of the value just happen once?)